File: src/languages/kurmanji/kurmanji.ts
/**
 * Kurmanji / Northern Kurdish (kmr) phonemizer — Iranian, the Latin (Hawar) alphabet, canonical IPA, espeak-
 * independent. A near-phonemic left-to-right scan (the digraph ⟨xw⟩→[xʷ], then single letters) + final-syllable
 * stress (Kurmanji default; unwritten, folded by the eval). Signature: ⟨c⟩→d͡ʒ / ⟨ç⟩→t͡ʃ, ⟨j⟩→ʒ, ⟨ş⟩→ʃ, ⟨q⟩→q,
 * ⟨x⟩→x; long a/ê/î/o/û vs short e/i/u. Aspiration/pharyngealisation are allophonic and not emitted. text()
 * tokenizes words / numbers / punctuation. See docs/investigations/kmr_native_bringup_investigation.md.
 */
import type { Phonemizer } from "../../registry.ts";
import { assembleClauses } from "../../core/clauses.ts";
import { numberToWords } from "./numbers.ts";
import { MANIFEST } from "./manifest.ts";

const DIGRAPHS = MANIFEST.digraphs;
const VOWELS = MANIFEST.vowels;
const CONS = MANIFEST.consonants;
const CLAUSE_MARK = MANIFEST.clausePunctuation;

interface Seg {
    ph: string;
    v: boolean;
}

/** Scan a lowercased Kurmanji word into IPA segments (xw digraph, then single letters). */
function toSegments(word: string): Seg[] {
    const w = word.toLowerCase();
    const segs: Seg[] = [];
    for (let i = 0; i < w.length; ) {
        const two = w.slice(i, i + 2);
        if (DIGRAPHS[two]) {
            segs.push({ ph: DIGRAPHS[two]!, v: false });
            i += 2;
            continue;
        }
        const c = w[i]!;
        if (VOWELS[c] !== undefined) segs.push({ ph: VOWELS[c]!, v: true });
        else if (CONS[c] !== undefined) segs.push({ ph: CONS[c]!, v: false });
        i++; // unknown char (punctuation) → skip
    }
    // Nasal place assimilation: /n/ → [ŋ] before a velar stop k/ɡ (bang→bɑːŋɡ, aheng→ɑːhɛŋɡ).
    for (let k = 0; k < segs.length - 1; k++)
        if (segs[k]!.ph === "n" && /^[kɡ]/.test(segs[k + 1]!.ph)) segs[k]!.ph = "ŋ";
    return segs;
}

/** One Kurmanji word → canonical IPA with final-syllable stress (before the last vowel nucleus). */
export function phonemizeWord(word: string): string {
    const segs = toSegments(word);
    const nuclei = segs.map((s, i) => (s.v ? i : -1)).filter((i) => i >= 0);
    if (nuclei.length === 0) return segs.map((s) => s.ph).join("");
    const stressIdx = nuclei[nuclei.length - 1]!; // final-syllable default
    let out = "";
    for (let i = 0; i < segs.length; i++) {
        if (i === stressIdx) out += "ˈ";
        out += segs[i]!.ph;
    }
    return out;
}

// A word (Kurmanji Hawar letters incl. diacritics ê î û ç ş, the dotted ğ ẍ ḧ, and the apostrophe ayn) / number /
// punctuation token. All three dotted letters that appear in the consonant map must be here or text() drops them.
const TOKEN = /([a-zêîûçşğẍḧ']+)|(\d+)|([.!?…,;:])/giu;

class KurmanjiPhonemizer implements Phonemizer {
    text(input: string): string {
        return assembleClauses(input, TOKEN, (m, sink) => {
            if (m[1]) sink.emit(phonemizeWord(m[1]));
            else if (m[2])
                for (const wd of numberToWords(Number(m[2])).split(" ")) sink.emit(phonemizeWord(wd));
            else if (m[3]) {
                const mk = CLAUSE_MARK[m[3]];
                if (mk) sink.pause(mk);
            }
        });
    }
}

/** Build the Kurmanji phonemizer (near-phonemic g2p + final-syllable stress). */
export function createKurmanji(): Phonemizer {
    return new KurmanjiPhonemizer();
}

File: csharp/Vernacula.Phonemizer/Languages/Kurmanji/Kurmanji.cs
/**
 * Kurmanji / Northern Kurdish (kmr) phonemizer — Iranian, the Latin (Hawar) alphabet, canonical IPA.
 * Near-phonemic left-to-right scan (the ⟨xw⟩ digraph, then single letters) + final-syllable stress.
 * Ported from src/languages/kurmanji/kurmanji.ts — see that file for the corpus evidence.
 */
using Vernacula.Phonemizer.Core;

namespace Vernacula.Phonemizer.Languages.Kurmanji;

public static class KurmanjiPhonemizer
{
    private static IReadOnlyDictionary<string, string> DIGRAPHS => Manifest.MANIFEST.Digraphs;
    private static IReadOnlyDictionary<string, string> VOWELS => Manifest.MANIFEST.Vowels;
    private static IReadOnlyDictionary<string, string> CONS => Manifest.MANIFEST.Consonants;
    private static IReadOnlyDictionary<string, string> CLAUSE_MARK => Manifest.MANIFEST.ClausePunctuation;

    private readonly record struct Seg(string Ph, bool V);

    private static readonly JsRe VELAR = JsRegex.Compile("^[kɡ]", "u");

    /** Scan a lowercased Kurmanji word into IPA segments (xw digraph, then single letters). */
    private static List<Seg> ToSegments(string word)
    {
        var w = Js.ToLowerCase(word);
        var segs = new List<Seg>();
        for (var i = 0; i < w.Length;)
        {
            var two = w[i..Math.Min(i + 2, w.Length)];
            if (DIGRAPHS.TryGetValue(two, out var dg))
            {
                segs.Add(new Seg(dg, false));
                i += 2;
                continue;
            }
            var c = w[i].ToString();
            if (VOWELS.TryGetValue(c, out var v)) segs.Add(new Seg(v, true));
            else if (CONS.TryGetValue(c, out var k)) segs.Add(new Seg(k, false));
            i++; // unknown char (punctuation) → skip
        }
        // Nasal place assimilation: /n/ → [ŋ] before a velar stop k/ɡ.
        for (var k = 0; k < segs.Count - 1; k++)
            if (segs[k].Ph == "n" && VELAR.IsMatch(segs[k + 1].Ph)) segs[k] = segs[k] with { Ph = "ŋ" };
        return segs;
    }

    /** One Kurmanji word → canonical IPA with final-syllable stress (before the last vowel nucleus). */
    public static string PhonemizeWord(string word)
    {
        var segs = ToSegments(word);
        var nuclei = segs.Select((s, i) => s.V ? i : -1).Where(i => i >= 0).ToList();
        if (nuclei.Count == 0) return string.Concat(segs.Select(s => s.Ph));
        var stressIdx = nuclei[^1]; // final-syllable default
        var outp = "";
        for (var i = 0; i < segs.Count; i++)
        {
            if (i == stressIdx) outp += "ˈ";
            outp += segs[i].Ph;
        }
        return outp;
    }

    private static readonly JsRe TOKEN = JsRegex.Compile(
        $"({HostWord.HostWordRun(new[] { "Latin" }, "'")})|(\\d+)|([.!?…,;:])", "giu");

    /** This language's OWN inventory — a token this class REJECTS carries a letter Kurmanji does not use. */
    private const string NATIVE_CLASS = "[a-zêîûçşğẍḧ']";
    private static readonly Func<string, string> Nat = HostWord.MakeNativiser(NATIVE_CLASS, "iu");

    private sealed class Engine : ILanguage
    {
        public string Text(string input)
        {
            return Clauses.AssembleClauses(Normalize.NormalizeKurmanji(input), TOKEN, (m, sink) =>
            {
                if (m.Groups[1].Success && m.Groups[1].Value.Length > 0) sink.Emit(PhonemizeWord(Nat(m.Groups[1].Value)));
                else if (m.Groups[2].Success && m.Groups[2].Value.Length > 0)
                {
                    foreach (var wd in KurmanjiNumbers.NumberToWords(Js.Number(m.Groups[2].Value), m.Groups[2].Value).Split(' '))
                        sink.Emit(PhonemizeWord(wd));
                }
                else if (m.Groups[3].Success && m.Groups[3].Value.Length > 0)
                {
                    if (CLAUSE_MARK.TryGetValue(m.Groups[3].Value, out var mk) && mk.Length > 0) sink.Pause(mk);
                }
            });
        }
    }

    /** Build the Kurmanji phonemizer (near-phonemic g2p + final-syllable stress). */
    public static ILanguage CreateKurmanji() => new Engine();

    internal static void RegisterSelf() => Registry.Register("kurmanji", CreateKurmanji);
}
